File: src/common/web/data/exporters/ProjectExporterDescriptor.ts
import { Type } from "class-transformer";

import { type ProjectFeatureID } from "../entities/project/features/ProjectFeature";

/**
 * The exporter ID type.
 */
export type ProjectExporterID = string;

/**
 * Flags specifying extended capabilities of the exporter.
 */
export const enum ProjectExporterCapabilities {
    None = 0,
    AutoExport = 0x0001
}

/**
 * Describes a project exporter. This class is used to easily transfer information about an exporter.
 *
 * @param exporter_id - The global exporter ID.
 * @param name - The display name.
 * @param description - The exporter's description.
 * @param extension - The extension of exported files.
 * @param scope - The scope where the exporter applies; if empty, it applies to the overall project.
 * @param defaultScope - A default scope when exporting if none is given.
 * @param defaultFilename - A default filename used when none is given.
 */
export class ProjectExporterDescriptor {
    public readonly exporter_id: ProjectExporterID;

    public readonly name: string;
    public readonly description: string;
    public readonly extension: string;

    // @ts-ignore
    @Type(() => String)
    public readonly scope: ProjectFeatureID[];

    public readonly capabilities: ProjectExporterCapabilities;

    public readonly default_scope: ProjectFeatureID | undefined;
    public readonly default_filename: string;

    public constructor(
        exporterID: ProjectExporterID,
        name: string,
        description: string,
        extension: string,
        scope: ProjectFeatureID[],
        capabilities: ProjectExporterCapabilities = ProjectExporterCapabilities.None,
        defaultScope: ProjectFeatureID | undefined = undefined,
        defaultFilename: string = ""
    ) {
        this.exporter_id = exporterID;

        this.name = name;
        this.description = description;
        this.extension = extension;

        this.scope = scope;

        this.capabilities = capabilities;

        this.default_scope = defaultScope;
        this.default_filename = defaultFilename;
    }
}

File: src/common/py/data/exporters/project_exporter_descriptor.py
import typing
from dataclasses import dataclass, field
from enum import IntFlag

from dataclasses_json.api import dataclass_json

from ..entities.project.features import ProjectFeatureID

ProjectExporterID = str
ProjectExporterScope = typing.List[ProjectFeatureID]


@dataclass_json
@dataclass(frozen=True, kw_only=True)
class ProjectExporterDescriptor:
    """
    Describes a project exporter. This class is used to easily transfer information about an exporter.

    Attributes:
        exporter_id: The global exporter ID.
        name: The display name.
        description: The exporter's description.
        extension: The extension of exported files.
        scope: The scope where the exporter applies; if empty, it applies to the overall project.
        capabilities: Extended exporter capabilities.
        default_scope: A default scope when exporting if none is given.
        default_filename: A default filename used when none is given.
    """

    class Capabilities(IntFlag):
        """
        Flags specifying extended capabilities of the exporter.
        """

        NONE = 0
        AUTO_EXPORT = 0x0001

    exporter_id: ProjectExporterID

    name: str
    description: str
    extension: str

    scope: typing.List[ProjectFeatureID] = field(default_factory=list)

    capabilities: Capabilities

    default_scope: ProjectFeatureID | None = None
    default_filename: str = ""
